### PR TITLE
Refactors for Gateway Server

### PR DESCRIPTION
**Pull request by briantigerchow**
This PR is comprised of preparatory refactors for an upcoming IPFS gateway-server binary.

* improves `corehttp` interface
* renames `corerepo` directory
* adds an `fsrepo.BestKnownPath()` method to handle order of precedence between the default FSRepo path and the path provided using the IPFS_PATH env variable
* provides `blockstore.AllKeys(ctx)` method and renames parameterized one to `AllKeysRange(ctx, 0, 0)`
* renames `corerepo.GarbageCollectAsync` and adds a synchronous version
* integrates eventlog into the `fsrepo` package so eventlogging works wherever `fsrepo`'s are used
* fixes a bug in the `eventlog` begin output

@jbenet @whyrusleeping

### CODE REVIEW

File: core/corerepo/gc.go [hunk 1]
```diff
@@ -14,9 +14,27 @@ type KeyRemoved struct {
...
+	for k := range keychan { // rely on AllKeysChan to close chan
+		if !n.Pinning.IsPinned(k) {
+			err := n.Blockstore.DeleteBlock(k)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
```
Comment (jbenet) [inline core/corerepo/gc.go]: im worried that the implementation of `GarbageCollect` and `GarbageCollectAsync` may drift. thoughts?

**Comment by jbenet** [comment]
LGTM RTM

File: core/corerepo/gc.go (hunk 1 above)
Comment (briantigerchow) [inline core/corerepo/gc.go]: The meat of the operations is limited to 3 statements. The rest is boilerplate async. It seems that sharing more code would make the synchronous version more complex and the asynchronous one less direct.

```Go
keychan, err := n.Blockstore.AllKeysChan(ctx)
if !n.Pinning.IsPinned(k) {
err := n.Blockstore.DeleteBlock(k)
```

Comment (jbenet) [inline core/corerepo/gc.go]: sgtm

**Pull request merged**